sstables/sstable_set.cc

```diff
@@ -61,94 +61,422 @@ std::ostream& operator<<(std::ostream& os, const sstables::sstable_run& run) {
     return os;
 }
 
-sstable_set::sstable_set(std::unique_ptr<sstable_set_impl> impl, schema_ptr s)
-        : _impl(std::move(impl))
-        , _schema(std::move(s))
-        , _all(make_lw_shared<sstable_list>()) {
-    // assumes impl contains no sstables
+versioned_sstable_set_data::versioned_sstable_set_data(std::unique_ptr<sstable_set_impl> impl) 
+    : _impl(std::move(impl)) {
 }
 
-sstable_set::sstable_set(const sstable_set& x)
-        : _impl(x._impl->clone())
-        , _schema(x._schema)
-        , _all(make_lw_shared<sstable_list>(*x._all))
-        , _all_runs(x._all_runs) {
+void versioned_sstable_set_data::insert(shared_sstable sst) {
+    if (_times_added.contains(sst)) {
+        _times_added.at(sst)++;
+        return;
+    }
+    try {
+        _impl->insert(sst);
+        _all_runs[sst->run_identifier()].insert(sst);
+        _all.insert(sst);
+        _times_added[sst] = 1;
+    } catch (...) {
+        _impl->erase(sst);
+        if (_all_runs.contains(sst->run_identifier())) {
+            _all_runs.at(sst->run_identifier()).erase(sst);
+        }
+        _all.erase(sst);
+        _times_added.erase(sst);
+        throw;
+    }
+}
+
+std::vector<shared_sstable> versioned_sstable_set_data::select(const dht::partition_range& range) const {
+    return _impl->select(range);
+}
+
+std::unordered_set<shared_sstable> versioned_sstable_set_data::select_by_run_id(utils::UUID run_id) const {
+    return _all_runs.at(run_id);
 }
 
-sstable_set::sstable_set(sstable_set&&) noexcept = default;
+const std::set<shared_sstable>& versioned_sstable_set_data::all() const {
+    return _all;
+}
 
-sstable_set&
-sstable_set::operator=(const sstable_set& x) {
-    if (this != &x) {
-        auto tmp = sstable_set(x);
-        *this = std::move(tmp);
+void versioned_sstable_set_data::remove(shared_sstable sst) {
+    if (--_times_added[sst] == 0) {
```

Comment:
look for `sst` in `_times_added` first and assert it's there, otherwise
you'd create an entry and decrement it to be -1.

Reply:
I think I should have used `at()` instead of brackets here. `sst` must be in `_times_added` because we're only calling `remove` on sstables from `_added` and `_erased` and we're assuring that `sst` is in `_times_added` when inserting to those two sets.